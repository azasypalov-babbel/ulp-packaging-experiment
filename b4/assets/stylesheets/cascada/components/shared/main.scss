@use "sass:math";
// Layout

%responsive-container {
  margin: 0 auto;
  width: 90%;

  @include media(medium) {
    max-width: 940px;
  }
  @include media(large) {
    // fixed width for large and beyond
    max-width: none;
    width: 940px;
  }
  @include media(extra-large) {
    width: 1170px;
  }
}

// Typography

%trainer-page-text {
  font-size: math.div(16, 15) * 1em;
  margin-bottom: rem(24px);
  text-align: center;
}

%centered-columns {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

// UI

%bordered-box {
  border-radius: $border-radius;
  background-color: $white;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16),
              0 0 2px 0 rgba(0, 0, 0, 0.12),
              inset 1px 1px 0 0 rgba(0, 0, 0, 0.08);
  color: $space-gray;
}

// Cascada override for Safari in iOS 10

.btn--primary,
.btn {
  display: inline-block;
}
