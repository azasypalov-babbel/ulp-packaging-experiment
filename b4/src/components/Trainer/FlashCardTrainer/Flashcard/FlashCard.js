import React from 'react';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { StyledFlashCard, StyledBackgroundImage, StyledContentArea, StyledHintArea } from './styles';
import { poseConfig } from './animations';
import withAnimations from '../../../shared/withAnimations';
import withKeypress from '../../../shared/withKeypress';
import withShortcutHint from '../../../shared/withShortcutHint';

export const FlashCard = React.forwardRef(({
  keyboardHintComponent,
  imageUrl,
  imageChildren,
  children,
  ...props
}, ref) => (
  <StyledFlashCard {...props} ref={ref}>
    {keyboardHintComponent && <StyledHintArea>{keyboardHintComponent}</StyledHintArea>}
    <StyledBackgroundImage src={imageUrl}>{imageChildren}</StyledBackgroundImage>
    <StyledContentArea>{children}</StyledContentArea>
  </StyledFlashCard>
));

FlashCard.displayName = 'FlashCard';

FlashCard.propTypes = {
  keyboardHintComponent: PropTypes.node,
  imageUrl: PropTypes.string,
  imageChildren: PropTypes.node,
  children: PropTypes.node,
  onClick: PropTypes.func
};

export default compose(
  withShortcutHint,
  withKeypress,
  withAnimations(poseConfig)
)(FlashCard);
