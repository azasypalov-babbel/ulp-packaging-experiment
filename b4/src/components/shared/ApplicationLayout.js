import styled, { css } from 'styled-components';
import { isiOS, isWebview } from '../../lib/features';

const fixNavigationBar = css`
  & .loy .loy-cascada-navbar {
    position: fixed;
    top: 0;
    padding-top: env(safe-area-inset-top);

    &:after {
      top: initial;
    }
  }
  & > main {
    margin-top: calc(${(props) => props.theme.size.navbar.xsmall.height} + env(safe-area-inset-top));
  }
`;

const disableSelection = css`
  user-select: none;
  -webkit-touch-callout: none;
`;

const ApplicationLayout = styled.div`
  display: flex;
  flex-direction: column;

  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 100vh; /* IE11 does not respect min-height for calculating flex */
  }

  min-height: 100vh;
  ${isiOS() && css`
    min-height: -webkit-fill-available; /* To reduce the page size if possible when software keyboard is open */
  `}

  ${isWebview() && fixNavigationBar};
  ${isWebview() && disableSelection};

  & > main {
    position: relative;
    flex: 1 0 auto; /* main should be least the height of it's content */

    display: flex;

    & > * {
      flex: 1 1 100%; /* make children 100% width, and shrink if screen width is too small */
    }
  }

  .hide-for-tiny {
    @media (max-width: 413px) {
      display: none;
    }
  }

  .fixed-for-small {
    @media (max-width: 767px) {
      position: fixed;
    }
  }

  .hide-for-small {
    @media (max-width: 767px) {
      display: none;
    }
  }
`;

export default ApplicationLayout;
