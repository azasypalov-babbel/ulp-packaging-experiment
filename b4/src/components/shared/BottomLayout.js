import React, { useEffect, useLayoutEffect, useState } from 'react';
import { getDisplayName } from '../getDisplayName';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPortal } from 'react-dom';
import useResizeObserver from './hooks/useResizeObserver';
import { getBottomInset, scrollIntoSafeArea, setBottomInset } from './scrollIntoSafeArea';
import { connect } from 'react-redux';

/**
 * This area is for any elements that are fixed to the bottom of the page
 * Give components `order` css property to organise them.
 */

const mountPointId = 'bottom-layout-mount-point';

export const StyledBottomLayout = styled.div`
  position: relative;

  @supports (position: sticky) {
    position: sticky;
    height: ${(props) => props.internalHeight}px;
  }

  bottom: 0;
  left: 0;
  right: 0;

  flex: 0 0 auto;
  z-index: 99;

  pointer-events: none;

  & > * {
    @supports (position: sticky) {
      position: absolute;
    }

    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;

    pointer-events: all;
  }
`;

export const bottomLayoutConfig = {
  enableScrollAdjustment: true,
  enableShrinking: false
};

const BottomLayoutComponent = ({ children, id }) => {
  const [internalHeight, setInternalHeight] = useState(0);

  const onResize = ({ height }) => {
    const isGrowing = height > internalHeight;

    if (isGrowing) {
      /**
       * InternalHeight is used to reserve space in the DOM.
       * It can only increase in height, to avoid making the page jump.
       */
      setInternalHeight(height);
    } else if (height !== 0 && bottomLayoutConfig.enableShrinking) {
      setInternalHeight(height);
    }

    const shouldAdjustScroll = height > getBottomInset();

    /**
     * Height of this element is used to calculate a safe area for scrolling
     * ie. so that content is not covered by anything in BottomLayout
     */
    setBottomInset(height);

    if (shouldAdjustScroll && bottomLayoutConfig.enableScrollAdjustment) {
      /**
       * Delay scroll until after next paint so that elements are in their final place
       */
      requestAnimationFrame(() => scrollIntoSafeArea());
    }
  };

  useEffect(() => {
    /**
     * Reset internal height when a the trainer changes or sequence starts or ends
     */
    setInternalHeight(0);
  }, [id]);

  const elementRef = useResizeObserver(onResize);

  return (
    <StyledBottomLayout internalHeight={internalHeight}>
      <div
        ref={elementRef}
        id={mountPointId}
      >
        {children}
      </div>
    </StyledBottomLayout>
  );
};

BottomLayoutComponent.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string
};

const mapStateToProps = ({ sequence }) => ({
  id: [sequence.currentTrainerIndex, sequence.started, sequence.completed].join()
});

export const BottomLayout = connect(mapStateToProps)(BottomLayoutComponent);

export const renderInBottomLayout = (WrappedComponent) => {
  const RenderInBottomLayout = (props) => {
    const [element, setElement] = useState(document.getElementById(mountPointId));

    useLayoutEffect(() => {
      const domNode = document.getElementById(mountPointId);
      if (!domNode) {
        // eslint-disable-next-line max-len
        console.warn(`DOM element with id ${mountPointId} not found. This is a requirement for a bottom layout to be properly rendered.`);
      }
      setElement(domNode);
    }, []);

    if (element === null) {
      return null;
    }

    return createPortal(
      <WrappedComponent {...props} />,
      element
    );
  };

  RenderInBottomLayout.displayName = `RenderInBottomLayout(${getDisplayName(WrappedComponent)})`;

  return RenderInBottomLayout;
};
