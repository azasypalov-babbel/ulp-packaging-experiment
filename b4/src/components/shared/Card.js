import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
