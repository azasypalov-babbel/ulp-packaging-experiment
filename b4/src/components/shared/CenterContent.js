import styled from 'styled-components';

const CenterContent = styled.div`
   display: flex;
   align-items: center;

   & > * {
      margin: auto;
   }
`;

export default CenterContent;
