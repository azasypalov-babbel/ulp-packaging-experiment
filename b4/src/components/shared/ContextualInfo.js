import React, { useEffect, useRef } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { markupStringToHTML } from '@lessonnine/babbel-markup-helper.js';

import Text from './Text';
import { scrollIntoSafeArea } from './scrollIntoSafeArea';

const StyledWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-direction: column;

  width: 100%;
  height: auto;

  background-color: ${(props) => props.theme.color.silver};

  border-radius: 0.75rem;

  padding: 1rem;
`;

const ContextualInfo = ({
  markupString,
  active,
  ...props
}) => {
  const ref = useRef();

  useEffect(() => {
    if (active) {
      scrollIntoSafeArea(ref.current);
    }
  }, [ref, active]);

  return (
    <StyledWrapper ref={ref} data-item-type={props.dataItemType}>
      <Text
        data-text-type="contextual-info"
        data-selector="non-interactive"
        fontFamily="fontMilliardBook"
        color="spaceGray"
        textAlign="center"
        dangerouslySetInnerHTML={{ __html: markupStringToHTML(markupString) }}
      />
    </StyledWrapper>
  );
};

ContextualInfo.propTypes = {
  active: PropTypes.bool,
  markupString: PropTypes.string.isRequired,
  dataItemType: PropTypes.string
};

export default ContextualInfo;
