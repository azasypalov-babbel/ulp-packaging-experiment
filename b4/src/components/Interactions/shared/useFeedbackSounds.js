import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
