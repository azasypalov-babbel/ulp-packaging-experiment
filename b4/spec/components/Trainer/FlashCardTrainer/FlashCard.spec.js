import React from 'react';
import { FlashCard } from '../../../../src/components/Trainer/FlashCardTrainer/Flashcard/FlashCard';
import { mountWithTheme } from '../../shared/themeMock';

describe('<FlashCard />', () => {
  test('it renders', () => {
    const wrapper = mountWithTheme(<FlashCard imageUrl="someUrl" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('with keyboardHintComponent', () => {
    test('it renders', () => {
      const keyboardHintComponent = <span>hint</span>;
      const wrapper = mountWithTheme(
        <FlashCard imageUrl="someUrl" keyboardHintComponent={keyboardHintComponent} />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
