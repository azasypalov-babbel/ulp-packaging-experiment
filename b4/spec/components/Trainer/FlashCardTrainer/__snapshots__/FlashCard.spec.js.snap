// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlashCard /> it renders 1`] = `
.c0 {
  box-sizing: border-box;
  overflow: hidden;
  background: #ffffff;
  border-radius: rem(4px);
  -webkit-transition: box-shadow 0.3s ease;
  transition: box-shadow 0.3s ease;
  box-shadow: 0 rem(2px) rem(20px) 0 transparentize(0.9,#565d64);
  outline: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: relative;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 1 45%;
  -ms-flex: 1 1 45%;
  flex: 1 1 45%;
  min-height: 14.5rem;
  background-image: url(someUrl);
  background-position: center center;
  background-size: cover;
  background-repeat: no-repeat;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 1 55%;
  -ms-flex: 1 1 55%;
  flex: 1 1 55%;
  padding: 1rem;
}

<FlashCard
  imageUrl="someUrl"
>
  <Styled(styled.div)>
    <div
      className="c0 c1"
    >
      <styled.div
        src="someUrl"
      >
        <div
          className="c2"
          src="someUrl"
        />
      </styled.div>
      <styled.div>
        <div
          className="c3"
        />
      </styled.div>
    </div>
  </Styled(styled.div)>
</FlashCard>
`;

exports[`<FlashCard /> with keyboardHintComponent it renders 1`] = `
.c0 {
  box-sizing: border-box;
  overflow: hidden;
  background: #ffffff;
  border-radius: rem(4px);
  -webkit-transition: box-shadow 0.3s ease;
  transition: box-shadow 0.3s ease;
  box-shadow: 0 rem(2px) rem(20px) 0 transparentize(0.9,#565d64);
  outline: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: relative;
}

.c2 {
  position: absolute;
  top: 1rem;
  left: 1rem;
  color: #284b7c;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 1 45%;
  -ms-flex: 1 1 45%;
  flex: 1 1 45%;
  min-height: 14.5rem;
  background-image: url(someUrl);
  background-position: center center;
  background-size: cover;
  background-repeat: no-repeat;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 1 55%;
  -ms-flex: 1 1 55%;
  flex: 1 1 55%;
  padding: 1rem;
}

<FlashCard
  imageUrl="someUrl"
  keyboardHintComponent={
    <span>
      hint
    </span>
  }
>
  <Styled(styled.div)>
    <div
      className="c0 c1"
    >
      <styled.div>
        <div
          className="c2"
        >
          <span>
            hint
          </span>
        </div>
      </styled.div>
      <styled.div
        src="someUrl"
      >
        <div
          className="c3"
          src="someUrl"
        />
      </styled.div>
      <styled.div>
        <div
          className="c4"
        />
      </styled.div>
    </div>
  </Styled(styled.div)>
</FlashCard>
`;
